Misc/aflytning-keyboard-warrior/index.md

```diff
@@ -0,0 +1,209 @@
+
+Unfortunately there's no descriptor packages sent to/from 3.14.3, so we can't tell what device it is.
+
+If we want to look at the actual keystroke being sent, we have to focus at one of two fields in Wireshark/tshark:
+* `usb.capdata`
+* `usbhid.data`
+
+Which is used, is entirely dependent on the device sending the packets.
+In our case, it seems like we're looking at `usbhid.data`, since a search for that provides 1870 packets, while `usb.capdata` gives 128.
+
+Lets get those packets extracted, that's where t-shark comes in:
+`tshark -r keyboard.pcapng -T fields -e usbhid.data > capdata.txt`
+
+This gives us a file containing all keystrokes from the keyboard, in this format:
+```
+200201000700000000000000000000
+200201000000000000000000000000
+200201000400000000000000000000
+200201000000000000000000000000
+200201000a00000000000000000000
+200201000000000000000000000000
+200201000500000000000000000000
+```
+
+Since this was my first time working with USB packets, I just started playing around with a python script, based on what I could find on the internet. Nothing matched perfectly (which I'm sure is on purpose from the creators), but I was able to figure out the structure pretty fast
+
+Manually manipulating the data  down to the fields that actually change between lines, I was left with this format:
+```
+020000
+020e00
+020000
+000400
+000800
+001500
+001508
+000800
+002c00
+020000
+...
+```
+
+The first byte is a modifier key. For example pressing shift, produces `020000`, whilst pressing another key with shift held down, produces `020e00`.
+
+The second byte is the character key being pressed, the bytes matched almost all tables I found online (eg. `04 == a`), so I'm guessing its a standard for american keyboards.
+
+The third character seems to be a repetition-marker. Every time its greater than 0, the key seems to be held down, and will be duplicated if transferred to the output.
+
+Based on that, I wrote this very [rudimentary Python script](scripts/decode_keyboard_v1.py):
+
+```python
+import os
+
+presses = []
+with open(os.path.relpath("capdata.txt"), "r") as f:
+    for line in f:
+        presses.append(line[0:-1])
+
+normalKeys = {"04":"a", "05":"b", "06":"c", "07":"d", "08":"e", "09":"f", "0a":"g", "0b":"h", "0c":"i", "0d":"j", "0e":"k", "0f":"l", "10":"m", "11":"n", "12":"o", "13":"p", "14":"q", "15":"r", "16":"s", "17":"t", "18":"u", "19":"v", "1a":"w", "1b":"x", "1c":"y", "1d":"z","1e":"1", "1f":"2", "20":"3", "21":"4", "22":"5", "23":"6","24":"7","25":"8","26":"9","27":"0","28":"<RET>","29":"<ESC>","2a":"<DEL>", "2b":"\t","2c":"<SPACE>","2d":"-","2e":"=","2f":"[","30":"]","31":"\\","32":"<NON>","33":";","34":"'","35":"<GA>","36":",","37":".","38":"/","39":"<CAP>","3a":"<F1>","3b":"<F2>", "3c":"<F3>","3d":"<F4>","3e":"<F5>","3f":"<F6>","40":"<F7>","41":"<F8>","42":"<F9>","43":"<F10>","44":"<F11>","45":"<F12>"}
```

Comment:
Holy smokes! :D

Reply:
Working code is working!
Don't always have time to make it pretty in a race ;)

This is probably a bit more legible: https://github.com/Nissen96/USB-HID-decoders/blob/main/keyboard_decode.py#L8